Scanning recipe to add dependency on detached plugin

CodexRaunak opened the pull request:
Issue #548
Stored all the details of the split plugins in an `enum`, with the name of their `classes`.
Scanning recipe matches any of the plugins used from the `DetachedPlugins`, if found then it checks if the `jenkins version > last core release` of that plugin which means now it has been detached, so it adds it in the `accumulator`.
Then in the visitor phase we `add the dependency` of the detached plugin.

### Testing done
Added the unit test for the recipe and tested it with `UpgradeToLatestJava8CoreVersion`.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

CodexRaunak commented:
Need a review on this PR, I am concerned about the enum part, is storing the class names of the plugins to match the plugin used or not, the correct way?

CodexRaunak commented:
![image](https://github.com/user-attachments/assets/847563dd-f14d-4042-abfc-d7a7f204cad0)

Also about this note, all the plugins between and including `matrix-auth` and `jaxb` have the incorrect `last core release` as mentioned.
So i need to change the enum `one version prior of the last core release.` As last core release defines the plugin is supported till that release.

jonesbusy commented:
> Need a review on this PR, I am concerned about the enum part, is storing the class names of the plugins to match the plugin used or not, the correct way?

Thanks for your PR. I would say if we can the package name. If we have some conflict (like LDAP) We could use class name.

I will need some time to review it but looks a very good start.

Can we have a test for every plugin?

jonesbusy commented:
> ![image](https://github.com/user-attachments/assets/847563dd-f14d-4042-abfc-d7a7f204cad0)
> 
> Also about this note, all the plugins between and including `matrix-auth` and `jaxb` have the incorrect `last core release` as mentioned.
> So i need to change the enum `one version prior of the last core release.` As last core release defines the plugin is supported till that release.
> 

I don't have more information about that. But I guess it's no so important since se will always modernizer to a much more recent versions

CodexRaunak commented:
> I don't have more information about that. But I guess it's no so important since se will always modernizer to a much more recent versions

Currently what I am doing is, I am checking if a plugin let's say `InstanceIdentity` is being used in a project. Then I am not directly adding it to the dependency, because maybe the jenkins version we are upgrading to may not have detached it. So that's why the last core release is maybe important to us, beacause it will help us check that we actually want to add it as a dependency or not.

If a plugin say `InstanceIdentity` is being used and its `lastCoreRelease` is `2.356`, so we only need to add it in dependency if the `jenkins version` is > `2.356`. If the `jenkinsVersion `is less than equal to `2.356 `don't add it in dependency as its not detached for that jenkins.

CodexRaunak commented:
I added the some test cases for every plugin, 
In the enum we are storing the `packageName `also, but if it has a conflict i.e more than one plugin use the same package then I added the core class names of that plugin.
Most of the plugins are overlapping so there package info is empty.

jonesbusy commented on plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/recipes/AddDetachedPluginDependency.java:
```diff
@@ -0,0 +1,116 @@
+                    document = (Xml.Document) new AddDependency(
+                                    plugin.getGroupId(),
+                                    plugin.getPluginId(),
+                                    "RELEASE",
```
Is the "latest.release" working?

CodexRaunak replied:
Works only on some of them , not all

jonesbusy commented:
Thanks! LGTM. We can still improve in the future if we discover some border case
